### ISSUE TITLE
Moving `DllGetClassObject` from ctypes to py2exe/comtypes

### ISSUE THREAD

**Issue by encukou** [py2exe/py2exe#217]
Hello, CPython developer here.

I recently found that `ctypes` imports and calls the third-party project `comtypes` -- something I'd like to stop doing, with as little disruption to users as possible. CPython is not really equipped to test this functionality... but maybe no one is?

We're kinda struggling to figure out what this is for, but, from [the discussions](https://github.com/python/cpython/issues/127369) it seems that the only practical way to use `DllGetClassObject` is by embedding Python in the way `py2exe` does. And py2exe has [some support for `comtypes` already](https://github.com/py2exe/py2exe/blob/e9c5f0e82a7a634346d87f8e8d2e9732935681a8/py2exe/boot_ctypes_com_server.py#L70). Could the support be moved to a more appropriate place?

The `DllCanUnloadNow` function is similar.

If there are any experts in this area here, your input would be welcome -- we don't want to break existing users, if there are any. We're discussing at https://github.com/python/cpython/issues/127369

**Comment by junkmd** [comment]
Hello, `comtypes` maintainer here.

I am concerned that `boot_ctypes_com_server` overwrites `comtypes.server.inprocserver._clsid_to_class`, and that there are classes and functions in `comtypes` that seem to exist solely for use by `py2exe`.

It is not an ideal situation for the standard library to hook into specific third-party APIs or for one third party to overwrite private attributes of another.

Additionally, based on https://github.com/py2exe/py2exe/issues/24, it seems that the functionality for implementing COM servers using `py2exe` does not work well.

Although I still don’t fully understand `boot_ctypes_com_server`, I believe there is room to discuss the specifications of `comtypes` and its future maintenance direction.

Furthermore, I welcome proposals to implement utility functions in `comtypes` that could serve as replacements for the current approach, making them usable not only with `py2exe` but also with other embedding tools.

**Comment by mhammond** [comment]
I don't know what the more fundamental questions are and that 10 year old issue doesn't really help me work out what they are. It was probably longer than 10 years since I ever used py2exe (but I do know that a number of COM objects were successfully shipped with it)

**Comment by junkmd** [comment]
These hooks and APIs broke a long time ago, but because so few people used them, the Python community may not have noticed, or bug reports may have been deprioritized, or the root cause of the bug remained unknown and the issue was stalled.

Using COM clients with `comtypes` or `pywin32` is a representative use case.
Additionally, freezing them with embedding tools is another very common use case.

However, implementing COM servers is extremely rare compared to client usage.
Implementing COM servers with Python and `comtypes` or `pywin32` is much rarer compared to implementations done with C# or C++.
Furthermore, implementing a frozen inproc COM server with Python embedding tools is an even rarer case.

In the Python 2.4 to 2.5 era, when Thomas Heller created `ctypes`, `comtypes`, and `py2exe`, it might be possible to implement frozen inproc COM servers by combining those tools.
However, as seen in enthought/comtypes#90 (also an issue from 10 years ago), this collaboration seemed to have broken by the time of Python 3.4.

Given this, it might be likely that no one is using these hooks at present effectively.

If that is the case, IMO, rather than making these hooks functional again, it would be better to design a modern protocol and enable the implementation of frozen in-proc servers based on it.
For the existing hooks, we could set a short deprecation period, and if no one requests their retention during that time, we can proceed to remove or make breaking changes to them.

**Comment by bennyrowland** [comment]
@albertosottile I know this isn't your area of expertise in Py2Exe and I don't expect you to necessarily fix this issue yourself, but I thought it might be a good idea to set down some thoughts about how it could be addressed, so that if anybody runs into this problem in the future they have some pointers on how to fix it.

When Py2Exe is freezing an inproc COM server into a DLL, it compiles [run_cypes_dll.c](https://github.com/py2exe/py2exe/blob/e9c5f0e82a7a634346d87f8e8d2e9732935681a8/source/run_ctypes_dll.c) file into the DLL, and also inserts a copy of [boot_ctypes_com_server.py](https://github.com/py2exe/py2exe/blob/e9c5f0e82a7a634346d87f8e8d2e9732935681a8/py2exe/boot_ctypes_com_server.py) as the boot script. When the COM machinery invokes this DLL the compiled code starts up a Python interpreter and runs the boot script.

The odd behaviour that is causing the problem is that the DLL needs to export several functions for the COM machinery to call into, and the way that 2 of these are obtained is by loading them out of the `_ctypes.pyd` DLL. To add extra complexity, the C functions in `_ctypes.pyd` then call into Python `ctypes` functions with the same name, which in turn directly forward the call on to `comtypes` functions.  But as far as we can work out, these functions are only in `ctypes` and `comtypes` in order for Py2Exe to be able to call them there, so it makes sense to move those functions out of `ctypes` and into Py2Exe.

However, I think that a better solution would be to simply move the `comtypes` [versions of these functions](https://github.com/enthought/comtypes/blob/46c34ec26a8ef315ae236e9bc852f22c5bec9355/comtypes/server/inprocserver.py#L112-L147) directly into [boot_ctypes_com_server.py](https://github.com/py2exe/py2exe/blob/e9c5f0e82a7a634346d87f8e8d2e9732935681a8/py2exe/boot_ctypes_com_server.py), and simply allow [run_cypes_dll.c](https://github.com/py2exe/py2exe/blob/e9c5f0e82a7a634346d87f8e8d2e9732935681a8/source/run_ctypes_dll.c) to invoke them from there. This is already done for the other important COM entry points `DllRegisterServer()` and `DllUnregisterServer()` ([C side](https://github.com/py2exe/py2exe/blob/e9c5f0e82a7a634346d87f8e8d2e9732935681a8/source/run_ctypes_dll.c#L273-L321) and [Python side](https://github.com/py2exe/py2exe/blob/e9c5f0e82a7a634346d87f8e8d2e9732935681a8/py2exe/boot_ctypes_com_server.py#L75-L90)). It should be pretty simple to move the `comtypes` functions across to the Py2Exe Python boot file and replicate the C side of the `DllRegisterServer()` implementations for `DllGetClassObject()` and `DllCanUnloadNow()`. Then `comtypes` and `ctypes` can get rid of their implementations without affecting anyone.

**Comment by AraHaan** [comment]
Speaking of how py2exe embeds the python interpreter I been working and experimenting on a way to make py2exe work with single file exe's for Python 3.12+.

How I plan to change how py2exe creates single file:

- have the python scripts read the version of cpython used and it's build architecture (is it x86, x64, or even ARM64)
  - Once this information is obtained download the embed zip file from python.org into a temp folder and extract every file **except python.exe and it's *._pth file**
  - Write a new ``*._pth`` file with the following paths: ``pythonXY.zip``, ``site-packages.zip``, and ``.`` (for loading the c extensions).
  - Once that is done, read the project's ``requirements.txt`` file (current working directory if it exists), then read their requirements to get a detailed list of all required dependencies. With this list we then run a compile step to transform all the ``*.py`` files to ``*.pyc`` files to zip up into ``site-packages.zip`` just like how ``pythonXY.zip`` is made in python/cpython today.
  - py2exe then copies a pre-built embed exe file that basically uses a frozen copy of ``memimport`` and ``zipextimporter`` that has been patched to also support Python 3.13+ (from [here](https://github.com/SeaHOH/memimport)). For 3.13+ this also has a linked in copy of ``audioop`` that is also ``PyImport_AppendInitTab``'d for compatibility with python programs that have not yet removed their dependency on ``audioop`` yet since it's removal from Python 3.13+. Packages like Discord.py need this as they plan to NEVER drop audioop and so they have someone practically forward port it for 3.13+ by compiling using the limited API with minor patches.
- After the above is done the then temp folder with all the required files to run the program standalone is then zipped up into a folder that then a pure C stub application then uses the ``miniz`` library (compiled into the exe) to read that then zip file from that stub exe's resource section. Once the zip file is read the entries are then extracted from it to a temp folder (that gets deleted upon termination of the real exe in that temp folder) and then ``CreateProcess`` is called on it.

What this means?
- For one I feel it would be a welcome cleanup of how py2exe can embed python programs. This also means that hopefully ctypes, and the other special cases py2exe accounts for today still works as is and so hopefully this way actually accomplishes that out of the box without special patches for such.

**Comment by albertosottile** [comment]
@junkmd 

> These hooks and APIs might be broken a long time ago, but because so few people used them, the Python community may not have noticed, or bug reports may have been deprioritized, or the root cause of the bug remained unknown and the issue was stalled.

> However, as seen in https://github.com/enthought/comtypes/issues/90 (also an issue from 10 years ago), this collaboration seemed to have broken by the time of Python 3.4. Given this, it might be likely that no one is using these hooks at present effectively.

I second all this. I am not aware of anyone that successfully managed to freeze a COM server with the code in this repository. As a matter of fact, I was already planning to delete `run_ctypes_dll.c` for a long time, because I also struggled in identifying a modern use case for it. I think this issue might now expedite the process.

@bennyrowland 

After what was discovered in https://github.com/python/cpython/issues/127369#issuecomment-2532227589, it seems to me these hooks are not solely for consumption from py2exe. Nevertheless, as I wrote before, I do not intend to support this feature anymore, unless a compelling use case shows up in the immediate future.

**Comment by albertosottile** [comment]
@emaimone @Avasam is there still demand/a use case for freezing a Python COM server in 2025?

**Comment by junkmd** [comment]
Sorry I'm late.

@albertosottile 

> I second all this. I am not aware of anyone that successfully managed to freeze a COM server with the code in this repository. As a matter of fact, I was already planning to delete `run_ctypes_dll.c` for a long time, because I also struggled in identifying a modern use case for it. I think this issue might now expedite the process.

I agree.
I believe more cautious discussion is needed regarding `ctypes`'s `DllGetClassObject` and `DllCanUnloadNow`.
On the other hand, as for freezing, it is highly likely that this functionality is not being used by anyone at all now.
However, I think it's necessary for both the `comtypes` and `py2exe` projects to coordinate the removal and release of the functionality.

If someone wants to freeze a COM server, I believe they'll need to create new entry points.
I hope that codebases will be more straightforward than reviving this dead codebases.

@emaimone @Avasam
What are your thoughts on this?

**Comment by Avasam** [comment]
> > I second all this. I am not aware of anyone that successfully managed to freeze a COM server with the code in this repository. As a matter of fact, I was already planning to delete `run_ctypes_dll.c` for a long time, because I also struggled in identifying a modern use case for it. I think this issue might now expedite the process.
> 
> [...] On the other hand, as for freezing, it is highly likely that this functionality is not being used by anyone at all now. However, I think it's necessary for both the `comtypes` and `py2exe` projects to coordinate the removal and release of the functionality.

> [@emaimone](https://github.com/emaimone) [@Avasam](https://github.com/Avasam) is there still demand/a use case for freezing a Python COM server in 2025?

I appreciate the ping, but I am not qualified when it comes to using Python COM servers in the real world. Mark H. (already in this conversation) would be. Maybe @geppi if we wanna page in all current pywin32 maintainers.

The most relevant issues I could find concerning pywin32, py2exe and COM servers are:
- https://github.com/mhammond/pywin32/issues/868
  - Links back to https://github.com/py2exe/py2exe/issues/24 and here
- https://github.com/mhammond/pywin32/issues/1132
  - Links to https://github.com/marcelotduarte/cx_Freeze/issues/333 which was also closed due to inactivity

And they seem to reflect most of @junkmd 's previous points, especially 
> However, as seen in https://github.com/enthought/comtypes/issues/90 (also an issue from 10 years ago), this collaboration seemed to have broken by the time of Python 3.4.

**Comment by albertosottile** [comment]
Is there any news or progress concerning this topic? I am asking because I am considering removing all the ctypes_dll / COM server code in the next release.

**Comment by junkmd** [comment]
There has been no progress on this issue for some time.
Personally, I haven't been able to allocate the resources to investigate this further recently.
On the other hand, since it is unclear whether there are any developers actively using this, I believe a phased approach—such as issuing a warning when this feature is used, rather than removing ctypes_dll or COM-related functionality immediately in the next release—would be the appropriate course of action.

**Comment by albertosottile** [comment]
Version 0.14.2.0 officially deprecates target `ctypes_comdll`. However, the feature has not been documented nor directly available without tampering some code in the `freeze` API, exclusive since version 0.14.1.0.

**Comment by encukou** [comment]
Thanks. I'll take it as a point for the “no one actually uses this” hypothesis.
